be/src/storage/compaction_manager.cpp

```diff
@@ -0,0 +1,146 @@
+// This file is licensed under the Elastic License 2.0. Copyright 2021-present, StarRocks Limited.
+
+#include "storage/compaction_manager.h"
+
+#include "storage/compaction_scheduler.h"
+#include "storage/data_dir.h"
+#include "util/thread.h"
+namespace starrocks {
+
+std::unique_ptr<CompactionManager> CompactionManager::_instance(new CompactionManager());
+
+CompactionManager* CompactionManager::instance() {
+    return _instance.get();
+}
+
+void CompactionManager::update_candidates(std::vector<CompactionCandidate> candidates) {
+    bool should_notify = false;
+    {
+        std::lock_guard lg(_candidates_mutex);
+        for (auto& candidate : candidates) {
+            size_t num = _compaction_candidates.erase(candidate);
+            should_notify |= num == 0;
+            _compaction_candidates.emplace(std::move(candidate));
+        }
+    }
+    VLOG(2) << "_compaction_candidates size:" << _compaction_candidates.size()
+            << ", to add candidates:" << candidates.size();
+    if (should_notify) {
+        VLOG(2) << "new compaction candidate added. notify scheduler";
+        _notify_schedulers();
+    }
+}
+
+void CompactionManager::insert_candidates(std::vector<CompactionCandidate> candidates) {
+    std::lock_guard lg(_candidates_mutex);
+    for (auto& candidate : candidates) {
+        _compaction_candidates.emplace(std::move(candidate));
+    }
+}
+
+CompactionCandidate CompactionManager::pick_candidate() {
+    std::lock_guard lg(_candidates_mutex);
+    if (_compaction_candidates.empty()) {
+        static CompactionCandidate invalid_candidate;
+        return invalid_candidate;
+    }
+
+    auto iter = _compaction_candidates.begin();
+    CompactionCandidate ret = *iter;
+    _compaction_candidates.erase(iter);
+    return ret;
+}
+
+void CompactionManager::update_tablet_async(TabletSharedPtr tablet, bool need_update_context, bool is_compaction) {
+    PriorityThreadPool::Task task;
+    task.work_function = [tablet, need_update_context, is_compaction, this] {
+        update_tablet(tablet, need_update_context, is_compaction);
+    };
+    bool ret = _update_candidate_pool.try_offer(task);
```

Comment:
there are very low possibility try_offer return false, but it doesn't mean fatal error. It maybe just large candidate with slow schedule since over load of machine. I just want to know is there any side-effect if we skip it when try_offer return false.

Reply:
the side-effect is that the tablet will not be compacted. I think adding a background patrol thread to check whether need to resubmit it will solve this problem. Is it ok?